compiler/codegen/OMRCodeGenerator.cpp

```diff
@@ -569,6 +569,89 @@ OMR::CodeGenerator::setUpForInstructionSelection()
 
   }
 
+/**
+ * \brief The uncommon-call-constant-node pass is an extra pre-codegen pass right
+ * before instruction selection. It aims to avoid the situation where
+ * a constant integer is made alive across calls and, thereby, lower the preserved register
+ * pressure caused due to this.
+ *
+ * \details
+ *
+ * \verbatim
+ * Take z/Architecture as an exmaple, we could run into the following situation where constant
+ * integer 16 is loaded to a preserved register GPR6, which is then loaded to GPR2 as call
+ * parameter. The GPR6 will be preserved by the callee and can incur extra register shuffling.
+ *
+ * LHI     GPR6,0x10
+ * LGR     GPR2,GPR6    ; LR=Reg_param
+ * ASSOCREGS
+ * Label L0036:
+ * LG      GPR14,#609 -216(GPR7)
+ * LGHI    GPR0,0xff28
+ * BASR    GPR14,GPR14         # Call "java/io/PrintStream.println(I)V"
+ * ASSOCREGS
+ * Label L0035:
+ * LGR     GPR2,GPR6    ; LR=Reg_move
+ * LGRL    GPR1,0x0000000000000000, targetAddr=0x000003FFE05A7C5A (offset=0x0000000000121C00)
+ * LLGF    GPR6, Shadow[<vft-symbol>] 0(GPR1) ; Throws Implicit Null Pointer Exception
+ * NILL    GPR6,0xff00
+ * ASSOCREGS
+ * Label L0041:
+ * LG      GPR14,#610 -216(GPR6)
+ * LGHI    GPR0,0xff28
+ * BASR    GPR14,GPR14         # Call "java/io/PrintStream.println(I)V"
+ *
+ * \endverbatim
+ *
+ * Loading integer constants as immediate values [i.e. LHI GPR2, 0x10] is preferred.
+ * The ranges for immedate value loading on different platforms are governed by the following queries:
+ *
+ * \verbatim
+ *   bool materializesLargeConstants() { return false;}
+ *   int64_t getLargestNegConstThatMustBeMaterialized() {return -32769;}
+ *   int64_t getSmallestPosConstThatMustBeMaterialized() {return 32768;}
+ * \endverbatim
+ *
+*/
+void
+OMR::CodeGenerator::uncommonCallConstNodes()
+   {
+   if(self()->comp()->getOption(TR_TraceCG))
+      {
+      traceMsg(self()->comp(), "Performing uncommon call constant nodes\n");
+      }
+
+   for (TR::TreeTop* tt = self()->comp()->getStartTree(); tt; tt = tt->getNextTreeTop())
+      {
+       TR::Node* node = tt->getNode();
+       if(node->getNumChildren() >= 1 &&
+               node->getFirstChild()->getOpCode().isFunctionCall())
+          {
+          TR::Node* callNode = node->getFirstChild();
+          for(uint32_t i = 0; i < callNode->getNumChildren(); ++i)
+             {
+             TR::Node* paramNode = callNode->getChild(i);
+             if(paramNode->getReferenceCount() > 1 &&
+                     paramNode->getOpCode().isLoadConst() &&
+                     !self()->isMaterialized(paramNode))
+                {
+                TR::Node* newConstNode = TR::Node::create(paramNode->getOpCodeValue(), 0);
+                newConstNode->setConstValue(paramNode->getConstValue());
+                callNode->setAndIncChild(i, newConstNode);
+                paramNode->decReferenceCount();
+
+                if(self()->comp()->getOption(TR_TraceCG))
```

Comment:
Convention says the trace should be before the transformation so that the log contains the line of the last transformation attempted in the event of a crash rather than leaving no trace of diagnostic output of what was about to happen.